File: algorithms_on_graphs/ExitFromMaze.hs
import           Control.Monad
import           Control.Monad.State
import qualified Data.Map            as M

main :: IO ()
main = do
    [nVertices, nEdges] <- fmap parseInts getLine
    edges <- fmap (map parseEdge) (replicateM nEdges getLine)
    let vs = toVertices nVertices edges
        exs = explore vs 1

    print (nVertices, nEdges)
    print edges
    print vs
    print exs

parseInts :: String -> [Int]
parseInts = fmap read . words

parseEdge :: String -> Edge
parseEdge = toPair . parseInts
    where toPair xs = (head xs, last xs)

type Key = Int
type Neighbour = Key
type Edge = (Key, Key)
type Vertices = M.Map Key [Neighbour]

toVertices :: Int -> [Edge] -> Vertices
toVertices nVertices = foldl step initMap
    where initMap = M.fromList [(k, []) | k <- [1..nVertices]]
          step vs (u, v) = M.adjust (v:) u . M.adjust (u:) v $ vs

type VisitedState = M.Map Key Bool

initialState :: Int -> VisitedState
initialState nVertices = M.fromList [(k, False) | k <- [1..nVertices]]

explore :: Key -> Vertices -> State VisitedState [Key]
explore key vMap = do
    visitedMap <- gets (M.adjust (const True) key)
    let neighbours = vMap M.! key
        unvisited = filter (not . (visitedMap M.!)) neighbours
        nUnvisited = length unvisited
    if null unvisited
       then return [a]
       else do
           explore


{-
explore :: Vertices -> Key -> [Key]
explore vs = recExplore initExpMap vs
    where initExpMap = M.fromList [(k, False) | k <- [1..nVertices]]
          nVertices = M.size vs

recExplore :: M.Map Key Bool -> Vertices -> Key -> [Key]
recExplore expMap vs key =
    let neighbours = vs M.! key
        unvisited = filter (not . (expMap M.!)) neighbours
        key' = if null unvisited then key else head unvisited
     in if null unvisited
           then []
           else recExplore (M.adjust not key' expMap) vs key'
-}

{-
vs = M.fromList [(1,[4,2]),(2,[3,1]),(3,[4,2]),(4,[1,3])]
nVertices = M.size vs
initExpMap = M.fromList [(k, False) | k <- [1..nVertices]]

4 4
1 2
3 2
4 3
1 4

-}

File: algorithms_on_graphs/exit_from_maze.py
# python3

class Vertex(object):
    def __init__(self, key, neighbours=None):
        self.key = key
        self.neighbours = neighbours or []

    def __str__(self):
        return 'V({}){}'.format(self.key, self.neighbours)
    __repr__ = __str__

def parse_ints():
    return [int(x) for x in input().split(' ')]

def parse_vertices_from_input():
    n_vertices, n_edges = parse_ints()
    vertices = {i + 1: Vertex(i + 1) for i in range(n_vertices)}
    edges = [parse_ints() for _ in range(n_edges)]
    for u, v in edges:
        vertices[u].neighbours.append(v)
        vertices[v].neighbours.append(u)
    return vertices

def explore(vertices, key):
    visited_bools = {k: False for k in vertices}
    visited_vertices = []
    recursive_explore(visited_bools, visited_vertices, vertices, key)
    return visited_vertices

def recursive_explore(visited_bools, visited_vertices, vertices, key):
    visited_bools[key] = True
    visited_vertices.append(key)
    for neighbour in vertices[key].neighbours:
        if not visited_bools[neighbour]:
            recursive_explore(visited_bools, visited_vertices, vertices,
                              neighbour)

if __name__ == '__main__':
    vertices = parse_vertices_from_input()
    u, v = parse_ints()
    is_connected = v in explore(vertices, u)
    print(int(is_connected))

'''

4 4
1 2
3 2
4 3
1 4

4 2
1 2
3 2
1 4

'''
